File: constellation_planning/algorithms/simulated_annealing.py
# -*- coding: utf-8 -*-
"""模拟退火算法"""

import random
import math
from typing import List, Any
from .base import PlanningAlgorithm, Solution, AlgorithmConfig


class SimulatedAnnealing(PlanningAlgorithm):
    """模拟退火算法"""
    
    def __init__(
        self,
        config: AlgorithmConfig = None,
        initial_temp: float = 100.0,
        cooling_rate: float = 0.995,
        min_temp: float = 0.01
    ):
        super().__init__(config)
        self.initial_temp = initial_temp
        self.cooling_rate = cooling_rate
        self.min_temp = min_temp
    
    def get_name(self) -> str:
        return f"SA(T0={self.initial_temp}, α={self.cooling_rate})"
    
    def solve(
        self,
        observations: List[Any],
        satellites: List[Any],
        **kwargs
    ) -> Solution:
        self._start_timer()
        
        if self.config.random_seed:
            random.seed(self.config.random_seed)
        
        # 初始解
        current = self._generate_initial_solution(observations, satellites)
        self._update_best(current)
        
        temp = self.initial_temp
        
        for iteration in range(self.config.max_iterations):
            if self._is_time_exceeded() or temp < self.min_temp:
                break
            
            # 生成邻域解
            neighbor = self._generate_neighbor(current, observations, satellites)
            
            # 计算能量差
            delta = neighbor.objective_value - current.objective_value
            
            # 接受准则
            if delta > 0:
                current = neighbor
            else:
                prob = math.exp(delta / temp)
                if random.random() < prob:
                    current = neighbor
            
            self._update_best(current)
            self.history.append(self.best_solution.objective_value)
            
            # 降温
            temp *= self.cooling_rate
        
        return self.best_solution
    
    def _generate_initial_solution(
        self,
        observations: List[Any],
        satellites: List[Any]
    ) -> Solution:
        """生成初始解"""
        solution = Solution()
        for obs in observations:
            if random.random() > 0.5 and hasattr(obs, 'satellite_id'):
                solution.assignments[obs.id] = obs.satellite_id
        solution.objective_value = len(solution.assignments)
        return solution
    
    def _generate_neighbor(
        self,
        current: Solution,
        observations: List[Any],
        satellites: List[Any]
    ) -> Solution:
        """生成邻域解"""
        neighbor = current.copy()
        obs = random.choice(list(observations))
        
        if obs.id in neighbor.assignments:
            del neighbor.assignments[obs.id]
        else:
            if hasattr(obs, 'satellite_id'):
                neighbor.assignments[obs.id] = obs.satellite_id
        
        neighbor.objective_value = len(neighbor.assignments)
        return neighbor

File: matlab_version/+algorithms/SimulatedAnnealing.m
classdef SimulatedAnnealing <
    algorithms.PlanningAlgorithm %
        SIMULATEDANNEALING 模拟退火 MATLAB 实现

            properties InitialTemp CoolingRate MinTemp end

                methods function obj = SimulatedAnnealing(config) obj =
    obj @algorithms.PlanningAlgorithm(config);
obj.InitialTemp = 100.0;
obj.CoolingRate = 0.995;
obj.MinTemp = 0.01;

if isfield (config, 'InitialTemp')
  , obj.InitialTemp = config.InitialTemp;
end end

    function bestSol = solve(obj, observations, satellites) tic;

% 预处理 obsMap = containers.Map();
            for
              i = 1 : length(observations),
              obsMap(observations(i).Id) = observations(i);
            end obsIds = obsMap.keys;

            % 初始解 current = obj.generateRandomSolution(observations);
            currentScore = obj.evaluate(current, obsMap);
            current.Objective = currentScore;

            obj.BestSolution = current; % Value copy for struct?
            % Struct with Map: Map is handle. We need deep copy function.
            obj.BestSolution.Assignments = containers.Map(current.Assignments.keys, current.Assignments.values);

            temp = obj.InitialTemp;

            for
              iter =
                  1 : obj.Config.MaxIterations if toc > obj.Config.TimeLimit ||
                  temp < obj.MinTemp,
              break;
            end

                % 生成邻域 neighbor = current;
            neighbor.Assignments = containers.Map(current.Assignments.keys,
                                                  current.Assignments.values);

            obsId = obsIds{randi(length(obsIds))};
            if neighbor
              .Assignments.isKey(obsId) remove(neighbor.Assignments, obsId);
            else
              obs = obsMap(obsId);
            neighbor.Assignments(obsId) = obs.SatelliteId;
            end

                neighborScore = obj.evaluate(neighbor, obsMap);
            delta = neighborScore - currentScore;

            if delta
              > 0 || rand() < exp(delta / temp) current = neighbor;
            currentScore = neighborScore;

            if currentScore
              > obj.BestSolution.Objective obj.BestSolution.Objective =
                  currentScore;
            obj.BestSolution.Assignments = containers.Map(
                current.Assignments.keys, current.Assignments.values);
            end end

                obj.History(end + 1) = obj.BestSolution.Objective;
            temp = temp * obj.CoolingRate;
            end

                obj.Runtime = toc;
            bestSol = obj.BestSolution;
            end end

                methods(Access = private) function sol =
                    generateRandomSolution(obj, observations) sol.Assignments =
                        containers.Map();
            for
              i = 1 : length(observations) if rand () >
                      0.5 sol.Assignments(observations(i).Id) =
                      observations(i).SatelliteId;
            end end end

                function score = evaluate(obj, solution, obsMap) score = 0;
            keys = solution.Assignments.keys;
            for
              i = 1 : length(keys) if obsMap.isKey(keys{i}) score =
                  score + obsMap(keys{i}).Score;
            end end end end end
